### PR TITLE
Reenable asynclazy tests

### PR DESCRIPTION
**Pull request by jasonmalinowski**
Unskip tests for issue #204

When a request to AsyncLazy gets a cancellation, we asynchronously cancel the underlying Task, since any continuations might run inline and cause surprises. This meant that if cancellation happened during a call to GetValueAsync or GetValue where the underlying computation was synchronously completed, there was a race where the task that gets returned might or might not be cancelled. It didn't strictly matter -- since cancellation is best-effort it was fine by the contract. But to test an earlier bug, we forced it into the non-cancelling case by simply stopping up the thread pool so the asynchronous cancellation never happened. That approach was really evil, so we try something else.

Now, we adjust the tests to assert the slightly stronger statement that once a computation runs and returns a value, it's cached, even if the value is not handed out. (The implementation already did this, but we now assert it.) We also tweak the cancellation handling to cancel the returned task more aggressively. This avoids the race between async cancellation and synchronous value returning in the context of the unit test, so we can assert behavior better. In general there's no guarantee, but it's now guaranteed "enough" our test can test the other behaviors well.

### CODE REVIEW

**Comment by jasonmalinowski** [comment]
@jaredpar is now an expert in AsyncLazy.

**Comment by jaredpar** [comment]
Never trust the experts ...

File: src/Workspaces/CoreTest/UtilityTest/AsyncLazyTests.StopTheThreadPoolContext.cs [hunk 1]
```diff
@@ -1,53 +0,0 @@
...
-namespace Microsoft.CodeAnalysis.UnitTests
-{
-    public partial class AsyncLazyTests
-    {
-        /// <summary>
-        /// For some of the AsyncLazy tests, we want to see what happens if the thread pool is
-        /// really behind on processing tasks. This code blocks up the thread pool with pointless work to 
-        /// ensure nothing ever runs.
-        /// </summary>
-        private class StopTheThreadPoolContext : IDisposable
```
Comment (jaredpar) [inline src/Workspaces/CoreTest/UtilityTest/AsyncLazyTests.StopTheThreadPoolContext.cs]: So happy to see the red here ...

File: src/Test/Utilities/AssertEx.cs [hunk 2]
```diff
@@ -378,7 +378,7 @@ public static void Throws<T>(Action del, bool allowDerived = false)
                 Assert.Equal(typeof(T), type);
             }
 
-            Assert.True(false, "No exception was thrown.");
+            throw new Exception("No exception was thrown.");
```
Comment (jaredpar) [inline src/Test/Utilities/AssertEx.cs]: Why did this change?

**Comment by jaredpar** [comment]
:+1:

File: src/Test/Utilities/AssertEx.cs (hunk 2 above)
Comment (jasonmalinowski) [inline src/Test/Utilities/AssertEx.cs]: I wanted AssertEx.Throws to return the exception type a la xunit's regular Assert.Throws. The compiler's not smart enough to know that Assert.True will always throw, so this was the easiest way to shut it up.

**Pull request merged**